File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    
    
    set<vector<int>>ans;
    void backtrack(vector<int>&candidates,int idx,vector<int>total,int sum,int target){
    
        if(sum>target or idx>=candidates.size()){
                return;
        }                
        if(sum==target){
                ans.insert(total);
                return;
        }
        
        total.push_back(candidates[idx]);
        backtrack(candidates,idx,total,sum+candidates[idx],target);
        total.pop_back();
        backtrack(candidates,idx+1,total,sum,target);
        
        
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        
            vector<int>total;
            int sum=0;
            backtrack(candidates,0,total,sum,target);
        
            vector<vector<int>>a(ans.begin(),ans.end());
            return a;
        
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    
    def __init__(self):
        
        self.ans=[]
        
        
    def helper(self,candidates,index,target,total,arr):
        
        if index>=len(candidates) or total>target:
                return
            
        if target==total:
                self.ans.append(arr[:])
                print(self.ans)
                return
        
        total+=candidates[index]
        arr.append(candidates[index])
        self.helper(candidates,index,target,total,arr)
        total-=candidates[index]
        arr.pop(-1)
        self.helper(candidates,index+1,target,total,arr)
        
    
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        arr=[]
        self.helper(candidates,0,target,0,arr)
        return self.ans
        